File: lib/isbn.ts
/**
 * Normalização e validação de ISBN para os identificadores do schema.org.
 *
 * Motivo: em 2026-08-21 o Search Console apontou "Valor ISBN13 inválido para
 * `isbn`" no relatório de Listagens do comerciante ([WNC-10030322]). A causa
 * era emitir `livro.isbn` cru — o valor do banco ia para o JSON-LD sem
 * nenhuma checagem de formato ou dígito verificador.
 *
 * Medido em 2026-08-21 (PostgREST, livros com `is_publishable=true` e `isbn`
 * não nulo, n=9): 7 eram ISBN-13 válidos, 1 tinha 13 dígitos com **checksum
 * errado** (`pai-rico-pai-pobre` → `9788576849943`) e 1 era **ISBN-10**
 * (`industrial-economics-and-management-principles` → `8131803015`). Esses 2
 * eram o apontamento.
 *
 * O Google valida `isbn` das Listagens do comerciante como ISBN-13, então
 * ISBN-10 é convertido (prefixo `978` + dígito verificador recalculado) em vez
 * de descartado — é a conversão canônica, não uma heurística.
 */

function checksum13(twelveDigits: string): number {
  const soma = [...twelveDigits].reduce(
    (acc, digito, i) => acc + (i % 2 === 0 ? 1 : 3) * Number(digito),
    0,
  );
  return (10 - (soma % 10)) % 10;
}

function isbn10Valido(corpo: string): boolean {
  // ISBN-10 usa peso decrescente 10..1 em módulo 11, e o último dígito pode
  // ser `X` (valor 10).
  const soma = [...corpo].reduce((acc, c, i) => {
    const valor = c === "X" ? 10 : Number(c);
    return acc + valor * (10 - i);
  }, 0);
  return soma % 11 === 0;
}

/**
 * Devolve o ISBN-13 canônico (13 dígitos, sem hífen) ou `null` quando o valor
 * não é um ISBN válido. Nunca devolve string vazia — quem chama pode testar a
 * verdade do retorno direto.
 */
export function toIsbn13(bruto: string | null | undefined): string | null {
  if (!bruto) return null;

  const limpo = bruto.toUpperCase().replace(/[^0-9X]/g, "");

  if (limpo.length === 13) {
    if (!/^\d{13}$/.test(limpo)) return null; // `X` só existe em ISBN-10
    return checksum13(limpo.slice(0, 12)) === Number(limpo[12]) ? limpo : null;
  }

  if (limpo.length === 10) {
    if (!/^\d{9}[0-9X]$/.test(limpo)) return null;
    if (!isbn10Valido(limpo)) return null;
    const corpo = `978${limpo.slice(0, 9)}`;
    return `${corpo}${checksum13(corpo)}`;
  }

  return null;
}

File: scripts/core/isbn.py
# ============================================================
# ISBN — normalizacao e validacao
# Livraria Alexandria
#
# Espelha `lib/isbn.ts` do site. As duas implementacoes precisam concordar:
# o site so emite `isbn`/`gtin13` no JSON-LD quando o valor e um ISBN-13
# valido, e o pipeline so grava ISBN que passe pela mesma regra.
#
# Motivo: em 2026-08-21 o Search Console apontou "Valor ISBN13 invalido para
# `isbn`" (Listagens do comerciante, [WNC-10030322]). O #289 fez o site parar
# de propagar o valor ruim; este modulo ataca a origem — a ingestao gravava o
# `isbn` do seed sem checar formato nem digito verificador.
#
# Medido em 2026-08-21 (PostgREST, livros com is_publishable=true e isbn nao
# nulo, n=9): 7 ISBN-13 validos, 1 com 13 digitos e checksum ERRADO
# (`pai-rico-pai-pobre` -> 9788576849943) e 1 ISBN-10
# (`industrial-economics-and-management-principles` -> 8131803015).
# ============================================================

import re


def _checksum13(doze_digitos):
    """Digito verificador do ISBN-13: modulo 10 com pesos alternados 1 e 3."""
    soma = sum(
        (1 if i % 2 == 0 else 3) * int(d)
        for i, d in enumerate(doze_digitos)
    )
    return (10 - (soma % 10)) % 10


def _isbn10_valido(corpo):
    """ISBN-10: modulo 11 com pesos decrescentes 10..1; o ultimo digito pode
    ser `X`, que vale 10."""
    soma = sum(
        (10 if c == "X" else int(c)) * (10 - i)
        for i, c in enumerate(corpo)
    )
    return soma % 11 == 0


def normalize_isbn13(bruto):
    """Devolve o ISBN-13 canonico (13 digitos, sem hifen) ou None.

    - ISBN-13 com checksum correto: devolve normalizado (sem hifen/espaco).
    - ISBN-10 valido: converte (prefixo 978 + checksum recalculado). E a
      conversao canonica, nao heuristica.
    - Qualquer outra coisa (checksum errado, tamanho estranho, lixo): None.

    Nunca levanta excecao — entrada pode ser None, numero ou string suja.
    """
    if bruto is None:
        return None

    limpo = re.sub(r"[^0-9X]", "", str(bruto).upper())

    if len(limpo) == 13:
        if not limpo.isdigit():
            return None  # `X` so existe em ISBN-10
        return limpo if _checksum13(limpo[:12]) == int(limpo[12]) else None

    if len(limpo) == 10:
        if not re.fullmatch(r"[0-9]{9}[0-9X]", limpo):
            return None
        if not _isbn10_valido(limpo):
            return None
        corpo = "978" + limpo[:9]
        return corpo + str(_checksum13(corpo))

    return None
